ecosystem/typescript/sdk/src/transaction_builder/builder_utils.ts

```diff
@@ -0,0 +1,307 @@
+    const res: TypeTag[] = [];
+    if (this.tokens.length <= 0) {
+      bail("Invalid type tag.");
+    }
+
+    while (this.tokens[0][1] !== endToken) {
+      res.push(this.parseTypeTag());
+
+      if (this.tokens.length > 0 && this.tokens[0][1] === endToken) {
+        break;
+      }
+
+      this.consume(",");
+      if (this.tokens.length > 0 && this.tokens[0][1] === endToken && allowTraillingComma) {
+        break;
+      }
+
+      if (this.tokens.length <= 0) {
+        bail("Invalid type tag.");
+      }
+    }
+    return res;
+  }
+
+  parseTypeTag(): TypeTag {
+    if (this.tokens.length === 0) {
+      bail("Invalid type tag.");
+    }
+
+    // Pop left most element out
+    const [tokenTy, tokenVal] = this.tokens.shift();
+
+    if (tokenVal === "u8") {
+      return new TypeTagU8();
+    }
+    if (tokenVal === "u64") {
+      return new TypeTagU64();
+    }
+    if (tokenVal === "u128") {
+      return new TypeTagU128();
+    }
+    if (tokenVal === "bool") {
+      return new TypeTagBool();
+    }
+    if (tokenVal === "address") {
+      return new TypeTagAddress();
+    }
+    if (tokenVal === "vector") {
+      this.consume("<");
+      const res = this.parseTypeTag();
+      this.consume(">");
+      return new TypeTagVector(res);
+    }
+    if (tokenTy === "IDENT" && (tokenVal.startsWith("0x") || tokenVal.startsWith("0X"))) {
+      const address = tokenVal;
+      this.consume("::");
+      const [moduleTokenTy, module] = this.tokens.shift();
+      if (moduleTokenTy !== "IDENT") {
+        bail("Invalid type tag.");
+      }
+      this.consume("::");
+      const [nameTokenTy, name] = this.tokens.shift();
+      if (nameTokenTy !== "IDENT") {
+        bail("Invalid type tag.");
+      }
+
+      let tyTags: TypeTag[] = [];
+      // Check if the struct has ty args
+      if (this.tokens.length > 0 && this.tokens[0][1] === "<") {
+        this.consume("<");
+        tyTags = this.parseCommaList(">", true);
+        this.consume(">");
+      }
+
+      const structTag = new StructTag(
+        AccountAddress.fromHex(address),
+        new Identifier(module),
+        new Identifier(name),
+        tyTags,
+      );
+      return new TypeTagStruct(structTag);
+    }
+
+    throw new Error("Invalid type tag.");
+  }
+}
+
+export function serializeArg(argVal: any, argType: TypeTag, serializer: Serializer) {
+  if (argType instanceof TypeTagBool) {
+    assertType(argVal, "boolean");
+    serializer.serializeBool(argVal);
+    return;
+  }
+  if (argType instanceof TypeTagU8) {
+    assertType(argVal, "number");
+    serializer.serializeU8(argVal);
+    return;
+  }
+  if (argType instanceof TypeTagU64) {
+    assertType(argVal, ["number", "bigint"]);
+    serializer.serializeU64(argVal);
+    return;
+  }
+  if (argType instanceof TypeTagU128) {
+    assertType(argVal, ["number", "bigint"]);
+    serializer.serializeU128(argVal);
+    return;
+  }
```

Comment:
What happens if these numbers are negative or floating point?

Should you check the limits so that they're within u8, u64 etc?